[driver] Make --version show if assertions, etc. are enabled

porglezomp opened the pull request:
It's useful to have some significant build options visible in the version when investigating problems with a specific compiler artifact. This makes it easy to see if assertions, expensive checks, sanitizers, etc. are enabled when checking a compiler version.

Example output:
```
clang version 19.0.0git (git@github.com:apple/llvm-project.git b4db1d4e5ceadcd8c0e0c010b36cdc7ae68ba236)
Target: x86_64-apple-darwin24.0.0
Thread model: posix
InstalledDir: /Users/cassie/Development/apple-llvm-project/build/bin
Build configuration: +assertions, +asan, +ubsan
```

Discussion thread: https://discourse.llvm.org/t/rfc-printing-more-build-config-in-clang-version/78112

porglezomp committed:
[driver] Make --version show if assertions, etc. are enabled

It's useful to have some significant build options visible in the
version when investigating problems with a specific compiler artifact.
This makes it easy to see if assertions, expensive checks, sanitizers,
etc. are enabled when checking a compiler version.

porglezomp commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+  // behavior. Intended for identifying the source of issues when reproducing
+  // changes.
+  std::vector<std::string> BuildOptions = {
+#if !__OPTIMIZE__
+      "+unoptimized",
```
Question for reviewers: Is this a good enough mechanism for the "unoptimized" build case, or do we actually want to plumb the `CMAKE_BUILD_TYPE` through?

boomanaiden154 replied:
Building clang/LLVM with MSVC is a supported configuration and is tested with post-commit bots, so that is definitely an expected configuration. I'm not exactly sure what the best course of action is though. I don't think there's a universally supported preprocessor macro for whether or not optimizations are enabled.

`CMAKE_BUILD_TYPE`  I think would be the best bet. https://github.com/llvm/llvm-project/pull/72021 made it so that there are a fixed number of values (unless additional flags are set). That would also let the user differentiate between something like `RELEASE` which is `-O3` and `MINSIZEREL`, which I believe is `-Oz`. If there is a different value specified, just displaying something like custom I think should be fine.

I don't really have any opinions on this though and will leave to others to decide what the best course of action should be. Just trying to provide additional context.

jroelofs replied:
I think `+unoptimized` is good enough for me. `CMAKE_BUILD_TYPE` doesn't give as much signal as it suggests, since toolchain files can stick in whatever CXX flags they want per build type.

porglezomp replied:
I was also thinking about the fact that the build type doesn't directly correspond to optimized or not.
I should at least make it so it at least doesn't always display `+unoptimized` on MSVC though, I can fix that.

porglezomp replied:
It turns out `llvm/Support/CommandLine` has a definition that's approximately correct for MSVC and uses `__OPTIMIZE__` for gnu compatible compilers, I switched to using that.

porglezomp commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+  };
+  if (!BuildOptions.empty()) {
+    OS << "Build configuration: ";
+    bool FirstOption = true;
+    for (const auto &Option : BuildOptions) {
+      if (!FirstOption)
+        OS << ", ";
+      OS << Option;
+      FirstOption = false;
+    }
```
Is there an existing helper for printing comma-separated lists I could use here?

boomanaiden154 replied:
`llvm::interleaveComma` would work I think.

porglezomp commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+      "+dfsan",
+#endif
+  };
+  if (!BuildOptions.empty()) {
```
Do we want it to be hidden for the default config?
My opinion: yes—because this is helpful to llvm developers but not to compiler end-users.

jroelofs replied:
Making it opt-out, and having the release scripts disable it seems like a good idea to me. It's not actionable for end-users, but if one has to remember to turn it on (or even know it's an option), one might not see the benefit of it in day-to-day development.

porglezomp replied:
Added this as `LLVM_VERSION_PRINTER_SHOW_BUILD_CONFIG` which defaults on, is there anything I should do to update the release scripts here or is that for someone else to do?

porglezomp commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+#endif
+  };
+  if (!BuildOptions.empty()) {
+    OS << "Build configuration: ";
```
Is the output style here good? Do we like `Build configuration:`?
Do we want this placed here? I put it here because it's at the end of the current mandatory output, so it seemed least likely to break any bad machine parsing that exists right now.

jroelofs replied:
`s/configuration/config/`?

porglezomp reviewed (commented):
Added questions I'd like to see reviewers address here.

JDevlieghere commented:
Would it be worth doing this for the llvm tools like dsymutil and dwarfdump as well? They already [print](https://github.com/llvm/llvm-project/blob/1f01c580444ea2daef67f95ffc5fde2de5a37cec/llvm/lib/Support/CommandLine.cpp#L2548) "Optimized build (with assertions)" but this PR covers a few more things, is more legible and it would be nice have more uniformity between the two.

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+  // Print out build configuration options that impact the compiler's runtime
+  // behavior. Intended for identifying the source of issues when reproducing
+  // changes.
+  std::vector<std::string> BuildOptions = {
```
This can be `const StringRef BuildOptions[] = {`

porglezomp replied:
It seems `interleaveComma` doesn't work with plain arrays currently (because of using `.begin()` instead of `std::begin()`).

felipepiovezan replied:
Since these are all const char*, you can make this an `ArrayRef<StringRef>`

jroelofs replied:
https://github.com/llvm/llvm-project/pull/87669

porglezomp replied:
I moved this to a global variable to support the API, so I think `ArrayRef<StringRef>` ends up being the appropriate type in the end.

MaskRay commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -2002,6 +2002,44 @@ void Driver::PrintVersion(const Compilation &C, raw_ostream &OS) const {
+  };
+  if (!BuildOptions.empty()) {
+    OS << "Build configuration: ";
+    bool FirstOption = true;
```
This can be simplified with llvm::interleaveComma

porglezomp committed:
Address review feedback

porglezomp committed:
Move LLVM_IS_DEBUG_BUILD into Support/CommandLine.h

Allow us to reuse this logic in the feature printing.

porglezomp committed:
Revert "Move LLVM_IS_DEBUG_BUILD into Support/CommandLine.h"

This reverts commit 1213ba2f965a0577fc6504e967ae17346da4ef4f.

porglezomp commented:
Moved the `LLVM_IS_DEBUG_BUILD` into the `llvm/Support` headers to reuse it, but thinking more about it I think it would be better to move the printing logic into `llvm/Support/CommandLine`(?) instead, to make it easier to reuse in other tools as @JDevlieghere is suggesting.

porglezomp committed:
[Support] Expose config printing as llvm::cl::printBuildConfig

porglezomp committed:
Add a config flag to disable build config printing for vendors

Some compilers released to end-users will want to force-disable printing
the build config, since that's intended for compiler developers and not
really end-users. Add that flag, defaulting to on, so compiler
developers get the most benefit but release engineers can hide it.

porglezomp commented:
The result of those refactors I now have a new build option to support hiding the version, should I be touching the `utils/bazel` with this patch?

jroelofs commented:
> The result of those refactors I now have a new build option to support hiding the version, should I be touching the `utils/bazel` with this patch?

cc @nico

jroelofs commented on clang/test/Driver/version-build-config.test:
```diff
@@ -0,0 +1,6 @@
+
+# CHECK: clang version
+# When assertions are enabled, we should have a build configuration line that reflects that
+# CHECK: Build configuration: {{.*}}+assertions
```
```suggestion
# CHECK: Build config: {{.*}}+assertions
```

jroelofs commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -2734,6 +2734,40 @@ void cl::PrintHelpMessage(bool Hidden, bool Categorized) {
+void cl::printBuildConfig(raw_ostream &OS) {
+#if LLVM_VERSION_PRINTER_SHOW_BUILD_CONFIG
+  OS << "Build config: ";
+  llvm::interleaveComma(cl::CompilerBuildConfig, OS,
+                        [&OS](const StringRef &Option) { OS << Option; });
```
```suggestion
  llvm::interleaveComma(cl::CompilerBuildConfig, OS);
```

porglezomp committed:
Apply suggestions from code review

Co-authored-by: Jon Roelofs <jroelofs@gmail.com>

JDevlieghere reviewed (approved):
Thanks, this LGTM!

MaskRay commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -2734,6 +2734,39 @@ void cl::PrintHelpMessage(bool Hidden, bool Categorized) {
+#if __has_feature(address_sanitizer)
+    "+asan",
+#endif
+#if __has_feature(undefined_behavior_sanitizer)
```
sort sanitizers alphabetically?

You can add thread_sanitizer and `hwaddress_sanitizer` as well

porglezomp replied:
Good catch on TSan.
I followed the documented list of `__has_feature` that I could see, I don't see `hwaddress_sanitizer` in the list here: https://clang.llvm.org/docs/LanguageExtensions.html#extensions-for-dynamic-analysis

porglezomp commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -2734,6 +2734,39 @@ void cl::PrintHelpMessage(bool Hidden, bool Categorized) {
     CommonOptions->CategorizedHiddenPrinter.printHelp();
 }
 
+ArrayRef<StringRef> cl::CompilerBuildConfig = {
```
I footgunned here, this doesn't do lifetime extension so this array is a temporary despite being a global. Making this a function to additionally avoid having a static constructor.

porglezomp committed:
More review fixups

- Sort sanitizers by name
- Add some missing sanitizers
- Oops! Fix a lifetime issue with the compiler build config

porglezomp commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -2734,6 +2734,48 @@ void cl::PrintHelpMessage(bool Hidden, bool Categorized) {
 }
 
+ArrayRef<StringRef> cl::getCompilerBuildConfig() {
+  static const StringRef Config[] = {
```
This fails to build when Config is empty

porglezomp committed:
Fix empty array issue in getCompilerBuildConfig

porglezomp committed:
clang-format
